From 3f9a4791b76f2764a72f09861b3a12bad9a83f13 Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Thu, 11 Apr 2024 10:53:45 +0800
Subject: [PATCH 2/2] perf(socket): improve `sockopt_set_ipv6_membership`

```
sock:setoption('ipv6_add_membership', { multiaddr = 'ff02::1', interface = 0 })
sock:setoption('ipv6_drop_membership', { multiaddr = 'ff02::1', interface = 0 })
```

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 socket.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/socket.c b/socket.c
index cff24ca..7e0460b 100644
--- a/socket.c
+++ b/socket.c
@@ -725,13 +725,21 @@ static int sockopt_set_ip_membership(struct eco_socket *sock, lua_State *L, stru
 
 static int sockopt_set_ipv6_membership(struct eco_socket *sock, lua_State *L, struct sock_opt *o)
 {
-    const char *addr = luaL_checkstring(L, 3);
     struct ipv6_mreq mreq;
+    const char *multiaddr;
+
+    luaL_checktype(L, 3, LUA_TTABLE);
+
+    lua_getfield(L, 3, "multiaddr");
 
-    if (inet_pton(AF_INET6, addr, &mreq.ipv6mr_multiaddr) != 1)
-        luaL_argerror(L, 3, "not a valid IPv6 address");
+    multiaddr = lua_tostring(L, -1);
+
+    if (!multiaddr || inet_pton(AF_INET6, multiaddr, &mreq.ipv6mr_multiaddr) != 1)
+        luaL_argerror(L, 3, "multiaddr: not a valid IPv6 address");
+
+    lua_getfield(L, 3, "interface");
 
-    mreq.ipv6mr_interface = luaL_optinteger(L, 4, 0);
+    mreq.ipv6mr_interface = luaL_optinteger(L, -1, 0);
 
     return sockopt_set(sock, L, o, &mreq, sizeof(mreq));
 }
-- 
2.34.1

